airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/IntegrationRunner.java

```diff
@@ -26,59 +26,90 @@
 
 import com.fasterxml.jackson.databind.JsonNode;
 import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
 import io.airbyte.commons.io.IOs;
 import io.airbyte.commons.json.Jsons;
+import io.airbyte.config.State;
 import io.airbyte.protocol.models.AirbyteCatalog;
 import io.airbyte.protocol.models.AirbyteMessage;
 import io.airbyte.protocol.models.AirbyteMessage.Type;
+import io.airbyte.protocol.models.AirbyteRecordMessage;
 import java.nio.file.Path;
 import java.util.Optional;
 import java.util.Scanner;
 import java.util.function.Consumer;
+import java.util.stream.Stream;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+/**
+ * Accepts EITHER a destination or a source. Routes commands from the commandline to the appropriate
+ * methods on the integration. Keeps itself DRY for methods that are common between source and
+ * destination.
```

Comment:
curious how people feel about this approach. I could have done a lot of stuff with subclassing etc, but this ended up being the way that kept all of this logic in one place where it could be consumed all at once. this was the easiest version i could come up with. it's a little weird because it means there's a bit of null handling. will leave it there and see if you can follow it.